escrow/src/main.sw

```diff
@@ -1,257 +1,346 @@
+        let mut escrow = storage.escrows.get(identifier);
 
-        if let Identity::Address(address) = sender.unwrap() {
-            require(address == storage.buyer.address || address == storage.seller.address, Error::UnauthorizedUser);
+        require(escrow.state == State::Pending, StateError::StateNotPending);
+        require(!escrow.disputed, StateError::AlreadyDisputed);
+        require(msg_sender().unwrap() == escrow.buyer.address, UserError::Unauthorized);
+        require(escrow.buyer.asset.is_some() && 0 < escrow.buyer.deposited_amount, StateError::CannotDisputeBeforeDesposit);
 
-            if address == storage.buyer.address {
-                require(storage.buyer.deposited, Error::DepositRequired);
+        // Lock the escrow
+        escrow.disputed = true;
+        storage.escrows.insert(identifier, escrow);
 
-                storage.buyer.deposited = false;
-                storage.buyer.approved = false;
+        log(DisputeEvent {
+            identifier
+        });
+    }
 
-                transfer_to_output(storage.asset_amount, storage.asset, storage.buyer.address);
-            } else if address == storage.seller.address {
-                require(storage.seller.deposited, Error::DepositRequired);
+    #[storage(read, write)]fn propose_arbiter(arbiter: Arbiter, identifier: u64) {
+        // The assertions ensure that only the seller can propose a new arbiter and the arbiter
+        // cannot be the buyer / seller, the arbiter will be able to take a none-zero payment
 
-                storage.seller.deposited = false;
-                storage.seller.approved = false;
+        let escrow = storage.escrows.get(identifier);
 
-                transfer_to_output(storage.asset_amount, storage.asset, storage.seller.address);
-            }
-        } else {
-            revert(0);
-        };
+        require(escrow.state == State::Pending, StateError::StateNotPending);
+
+        let user = msg_sender().unwrap();
+
+        require(user == escrow.seller.address, UserError::Unauthorized);
+        require(arbiter.address != escrow.buyer.address, ArbiterInputError::CannotBeBuyer);
+        require(arbiter.address != escrow.seller.address, ArbiterInputError::CannotBeSeller);
+        require(0 < arbiter.fee_amount, ArbiterInputError::FeeCannotBeZero);
+        require(arbiter.fee_amount == msg_amount(), ArbiterInputError::FeeDoesNotMatchAmountSent);
+        require(arbiter.asset == msg_asset_id(), ArbiterInputError::AssetDoesNotMatch);
+
+        // If there is a previous proposal then we must transfer those funds back to the seller
+        let proposal = storage.arbiter_proposal.get(identifier);
+        if proposal.is_some() && 0 < proposal.unwrap().fee_amount {
+            transfer(proposal.unwrap().fee_amount, proposal.unwrap().asset, escrow.seller.address);
+        }
 
-        true
+        storage.arbiter_proposal.insert(identifier, Option::Some(arbiter));
```

Comment:
In Solidity we perform all state changes first to prevent re-entrency / use a modifier. In Sway, when transferring, that is unnecessary because the transfer does not allow re-entrency since it's not a standard contract call. It simply performs a transfer so there is no vector for re-entrency when transferring. That being said it's still a good idea to keep that pattern in mind.